Allow linking badge templates to regforms

tomasr8 opened the pull request:
Something that has been asked many times before is the ability to reference regform fields in the badge designer.
Currently you can use any regform as a source when printing badges and the list of available placeholders is limited to personal data since every regform has them.

The idea is to essentially allow linking templates to a specific regform. Once linked, you would be able reference any field from the linked regform.

Once linked, you have all registration data placeholders available in the badge designer and you can print the badges.

When printing badges/tickets, you can only select the linked regform (if any). The equivalent is true when configuring default ticket template in regforms (TicketsForm) - you can only select unlinked templates, or templates which are linked to the given regform.

This is what the UI for linking looks like currently:
![image](https://github.com/indico/indico/assets/8739637/d385fbd0-15dd-44d3-89a4-5f50aaab3357)

For not currently linked templates, there is a 'plus' icon which opens dropdown where you can select to link a regform (same as when printing badges). To unlink, you click on the red 'X'.

ThiefMaster commented:
For double-sided:

- Front/back side must be linked to the same regform (this implies both templates are linked to the event itself), or the backside must not be linked to anything
- Linking a template to a regform should not be allowed if it's the backside of another template, unless that template is linked to the same registration form

ThiefMaster commented on indico/modules/designer/models/templates.py:
```diff
@@ -39,6 +41,12 @@ class DesignerTemplate(db.Model):
         nullable=False
     )
+    registration_form_id = db.Column(
+        db.Integer,
```
not needed, `db.ForeignKey` looks up the referenced type automatically (which nobody knew early on so in most places we explicitly repeat it)

ThiefMaster commented on indico/modules/designer/models/templates.py:
```diff
@@ -76,6 +84,15 @@ class DesignerTemplate(db.Model):
+    regform = db.relationship(
+        'RegistrationForm',
+        lazy=True,
+        foreign_keys=registration_form_id,
```
is this needed? there's only one FK between those tables

ThiefMaster commented on indico/modules/designer/placeholders.py:
```diff
@@ -5,10 +5,12 @@
 
 from babel.numbers import format_currency
 from PIL import Image
+from speaklater import _LazyString
```
`from speaklater import is_lazy_string`

bpedersen2 commented:
Looks promising, if ready I can also try to review it in detail.

tomasr8 commented:
> Looks promising, if ready I can also try to review it in detail.

Hi! Yes, definitely feel free to give it a try :) It should be more or less working at this point, I mostly need to add tests.
I rebased it on the latest master and updated the migration so you just need `indico db upgrade`

duartegalvao commented on indico/modules/designer/templates/_list.html:
```diff
@@ -60,9 +68,43 @@
             <div class="thin toolbar right">
                 <div class="group">
+                    {% if not tpl.registration_form and linkable_forms[tpl] %}
+                        <button class="i-button icon-plus icon-only highlight text-color js-dropdown borderless"
```
What about using an icon more indicative to what this does... Was thinking either `icon-link` or `icon-clipboard` (which we use for regforms). What do you think?
![image](https://github.com/indico/indico/assets/27357203/c944c4b9-5c6d-46cc-8505-67634d2f9fbb)

tomasr8 replied:
I went with the feather icon (`icon-quill`) in the end since that's what we use in most places for regforms

duartegalvao commented on indico/modules/designer/templates/_list.html:
```diff
@@ -60,9 +68,43 @@
+                <a href="{{ url_for('event_registration.manage_regform', tpl.registration_form) }}">
+                    {{ tpl.registration_form.title }}
+                </a>
+                <button class="i-button icon-cross icon-only text-color borderless danger"
+                        title="{% trans %}Unlink registration form{% endtrans %}"
+                        data-href="{{ url_for('.unlink_regform', tpl) }}"
+                        data-method="PATCH"
+                        data-update="#template-list"
+                        data-confirm="{% trans title=tpl.registration_form.title -%}
+                                            Are you sure you want to unlink the registration form '{{ title }}'?
+                                            Any placeholders specific to that registration form will be removed
+                                            from the template.
+                                      {%- endtrans %}">
```
Maybe it would look better with a bit less `padding-left`?
![image](https://github.com/indico/indico/assets/27357203/bc6d4820-60b1-48ae-91d6-34fca346de7b)
(here I'm using `0.1em`)

duartegalvao commented on indico/modules/designer/placeholders.py:
```diff
@@ -33,15 +33,17 @@
+    'event': {'title': _('Event Data'), 'position': 1},
+    'fixed': {'title': _('Fixed Data'), 'position': 2},
+    'registrant': {'title': _('Common Registrant Data'), 'position': 3},
+    # Come up with a better name or merge with registrant data
+    'dynamic': {'title': _('Custom Registrant Data'), 'position': 4},
```
may be a bit wordy but something like `registrant_custom` would at least be more obvious

duartegalvao reviewed (commented):
seems to work well!

ThiefMaster commented on indico/modules/designer/models/templates.py:
```diff
@@ -152,6 +168,18 @@ def is_ticket(self):
+    def link_regform(self, regform):
+        """Link this template to a registration form."""
+        self.registration_form = regform
+        db.session.commit()
```
i'd rather not have commits in such places. we already commit at the end of a request if there's no error, and usually one does not expect some method on a model to commit early

tomasr8 commented on indico/modules/designer/__init__.py:
```diff
@@ -80,13 +80,20 @@ class PageLayout(RichIntEnum):
 
 @signals.core.get_placeholders.connect_via('designer-fields')
-def _get_notification_placeholders(sender, **kwargs):
+def _get_notification_placeholders(sender, regform=None, **kwargs):
```
Hmm, why is the method named this way? It has nothing to do with notifications :thinking:

ThiefMaster commented on indico/modules/designer/placeholders.py:
```diff
@@ -407,3 +408,65 @@ class FixedImagePlaceholder(DesignerPlaceholder):
+        arrival = format_date(friendly_data['arrival_date'])  # TODO: include locale
+        departure = format_date(friendly_data['departure_date'])
+        accommodation = friendly_data['choice']
+        return _('''\
+Arrival: {arrival} · Departure {departure} · Accommodation: {accommodation}
+''').format(arrival=arrival, departure=departure, accommodation=accommodation)
```
I'd rather use a neutral version that does not require i18n, e.g. `{accommodation} ({arrival} - {departure})`.
When someone really puts that field on a badge, they'll probably have to put static text around it anyway.

Because if no event language is set (few events do that) (and even then you'd need to force the event locale which I don't think we do anywhere when printing badges?), using the language of the person generating the badge - in particular if it's a manager - is not ideal. Image a Polish or Chinese user at CERN printing badges for some event happening here... ;)

ThiefMaster commented on indico/modules/designer/placeholders.py:
```diff
@@ -407,3 +408,65 @@ class FixedImagePlaceholder(DesignerPlaceholder):
         return Image.open(buf)
+
+
+class RegistrationFormFieldPlaceholder(DesignerPlaceholder):
```
It would be nice if this also took the (somewhat new) picture field into account.

Usually this is checked (in various places, also e.g. in the designer to use a square-shaped item) by looking at the class' `is_image`, but of course we cannot do that here because this placeholder can be either an image or nor depending on the field...

ThiefMaster commented on indico/modules/designer/controllers.py:
```diff
@@ -190,6 +203,7 @@ def clone_template(self, target=None):
         new_template = DesignerTemplate(title=title, type=self.template.type, **target_dict)
 
         data = deepcopy(self.template.data)
+        data = remove_regform_field_placeholders(data)
```
I'd rather clone it as linked. It's nice if you can have a backup or want to make just some minor changes!

ThiefMaster commented on indico/modules/designer/controllers.py:
```diff
@@ -340,6 +356,48 @@ def _process_POST(self):
+        if not regform:
+            raise ValueError('This template is not linked to any registration form.')
+
+        self.template.unlink_regform()
```
this is quite destructive, we should have a confirmation prompt in the UI for unlinking. or should we even allow unlinking as long as there are custom placeholders?

ThiefMaster replied:
ah hm, maybe i unconsciously clicked "yes" :D

anyway, i think preventing is better, i can't see a good reason why you'd add custom fields and then unlink... "least surprise for the user" is generally good. we all know how much users read confirmation prompts (\*cough\* event deletion \*cough\*)

tomasr8 commented on indico/migrations/versions/20240314_1430_4e32f4d5ebe4_add_registration_id_to_designer_templates.py:
```diff
@@ -0,0 +1,30 @@
+                               '(category_id IS NULL) OR (registration_form_id IS NULL)', schema='indico')
+
+
+def downgrade():
```
Should we delete custom fields from the template data when downgrading? Keeping the data around prevents the template from being modified unless you first delete all custom fields:
![image](https://github.com/indico/indico/assets/8739637/18c772a0-1749-499b-a46b-fb8f174186b4)

On the other hand if you ever want to upgrade afterwards you'll lose the data. If we keep the fields you can simply link to the right regform again and it'll work.

ThiefMaster replied:
let's keep it, it only breaks saving when you still have the data there, so not a big deal. downgrades are rare anyway, we just want to make sure we do not get a state where indico itself is broken or pages fail on load

ThiefMaster commented on indico/modules/designer/models/templates.py:
```diff
@@ -152,6 +168,22 @@ def is_ticket(self):
+    def unlink_regform(self):
+        """Unlink this template from a registration form."""
+        from indico.modules.designer.util import remove_regform_field_placeholders
+        self.data = remove_regform_field_placeholders(self.data)
```
Maybe this isn't needed anymore?

ThiefMaster commented on indico/modules/designer/placeholders.py:
```diff
@@ -422,6 +422,7 @@ class RegistrationFormFieldPlaceholder(DesignerPlaceholder):
     def __init__(self, field):
         self.field = field
         self.name = f'field-{field.id}'
+        self.is_image = self.field.input_type == 'picture'
```
Nice that this works. I was somehow thinking the signal checks it on the class itself and not on an instance of it. But with instances it's of course easy to handle...

ThiefMaster commented on indico/modules/designer/__init__.py:
```diff
@@ -80,13 +80,20 @@ class PageLayout(RichIntEnum):
+        from indico.modules.designer.placeholders import RegistrationFormFieldPlaceholder
+        from indico.modules.events.registration.models.items import RegistrationFormItemType
+        for field in regform.active_fields:
+            if field.type != RegistrationFormItemType.field_pd:      # Personal data fields are already included
+                yield RegistrationFormFieldPlaceholder(field=field)  # in the 'registrant' group
```
```suggestion
            # Personal data fields are already included in the 'registrant' group
            if field.type != RegistrationFormItemType.field_pd:
                yield RegistrationFormFieldPlaceholder(field=field)
```

not a fan of "block-style" comments ;)

ThiefMaster commented on indico/modules/designer/templates/_list.html:
```diff
@@ -1,3 +1,11 @@
+{% macro _render_link_regform_button(regform, tpl) %}
+    <a data-href="{{ url_for('designer.link_regform', regform, template_id=tpl.id) }}"
+       data-method="PATCH"
```
```suggestion
       data-method="POST"
```

ThiefMaster commented on indico/modules/designer/util.py:
```diff
@@ -98,3 +104,55 @@ def get_badge_format(tpl):
+
+    A template can be used if it is either not linked to anything or linked to the current registation form.
+    """
+    return DesignerTemplate.query.with_parent(regform.event).filter(
+        db.or_(
+            DesignerTemplate.registration_form_id.is_(None),
+            DesignerTemplate.registration_form == regform
+        )
+    ).all()
```
Our usual formatting for multiline queries

```suggestion
    return (
        DesignerTemplate.query
        .with_parent(regform.event)
        .filter(db.or_(DesignerTemplate.registration_form_id.is_(None),
                       DesignerTemplate.registration_form == regform))
        .all()
    )
```

ThiefMaster commented on indico/modules/designer/blueprint.py:
```diff
@@ -35,6 +36,10 @@ def view_func(**kwargs):
 _bp.add_url_rule('/category/<int:category_id>/manage/designer/<int:template_id>/toggle-default-badge',
                  'toggle_category_default_badge', RHToggleBadgeDefaultOnCategory, methods=('POST',))
 
+_bp.add_url_rule('/event/<int:event_id>/manage/designer/<int:template_id>/link/<int:reg_form_id>', 'link_regform',
+                 RHLinkDesignerTemplate, defaults={'object_type': 'event'}, methods=('PATCH',))
+_bp.add_url_rule('/event/<int:event_id>/manage/designer/<int:template_id>/unlink', 'unlink_regform',
+                 RHUnLinkDesignerTemplate, defaults={'object_type': 'event'}, methods=('PATCH',))
```
```suggestion
_bp.add_url_rule('/event/<int:event_id>/manage/designer/<int:template_id>/link/<int:reg_form_id>', 'link_regform',
                 RHLinkDesignerTemplate, defaults={'object_type': 'event'}, methods=('POST',))
_bp.add_url_rule('/event/<int:event_id>/manage/designer/<int:template_id>/unlink', 'unlink_regform',
                 RHUnLinkDesignerTemplate, defaults={'object_type': 'event'}, methods=('POST',))
```

"Non-RESTful" actions usually use POST. It's strange to use PATCH without a payload indicating what to update

ThiefMaster commented on indico/modules/designer/templates/_list.html:
```diff
@@ -60,9 +68,46 @@
+                        {% if unlinkable %}
+                            title="{% trans %}Unlink registration form{% endtrans %}"
+                            data-href="{{ url_for('.unlink_regform', tpl) }}"
+                            data-method="PATCH"
```
```suggestion
                            data-method="POST"
```
